src/libmeasurement_kit/ooni/whatsapp.cpp

```diff
@@ -0,0 +1,575 @@
+                     url.c_str());
+             }
+             *urls_tested += 1;
+             if (urls_count == *urls_tested) {
+                 cb(NoError());
+             }
+         };
+    };
+
+    for (auto url : urls) {
+        options["http/url"] = url;
+        http::Headers headers = constants::COMMON_CLIENT_HEADERS;
+        std::string body;
+        templates::http_request(entry, options, headers, body,
+                                http_cb(url), reactor, logger);
+    }
+
+    return;
+}
+
+void whatsapp(std::string input, Settings options,
+              Callback<Var<report::Entry>> callback,
+              Var<Reactor> reactor, Var<Logger> logger) {
+    std::vector<std::string>
+        WHATSAPP_REG_URLS = { "https://v.whatsapp.net/v2/register" };
+    std::vector<std::string>
+        WHATSAPP_WEB_URLS = { "https://web.whatsapp.com/",
+                              "http://web.whatsapp.com/" };
+    // XXX generate these with a range or something
+    std::vector<std::string>
+        WHATSAPP_ENDPOINT_HOSTNAMES = { "e1.whatsapp.net",
+//                                        "e2.whatsapp.net",
+//                                        "e3.whatsapp.net",
+//                                        "e4.whatsapp.net",
+//                                        "e5.whatsapp.net",
+//                                        "e6.whatsapp.net",
+//                                        "e7.whatsapp.net",
+//                                        "e8.whatsapp.net",
+//                                        "e9.whatsapp.net",
+//                                        "e10.whatsapp.net",
+//                                        "e11.whatsapp.net",
+//                                        "e12.whatsapp.net",
+//                                        "e13.whatsapp.net",
+//                                        "e14.whatsapp.net",
+//                                        "e15.whatsapp.net",
+                                        "e16.whatsapp.net" };
+
+    logger->info("starting whatsapp");
+    Var<Entry> entry(new Entry);
+
+//    // XXX pick smaller random subset of endpoints to test?
+//    http_many(WHATSAPP_REG_URLS, "registration_server", entry, [=](Error err) {
+//        logger->info("done testing whatsapp registration");
+//        http_many(WHATSAPP_WEB_URLS, "whatsapp_registration", entry, [=](Error err) {
+//            logger->info("done testing whatsapp web");
+//            dns_many(WHATSAPP_ENDPOINT_HOSTNAMES, entry,
+//                    [=](Error err, std::vector<std::string> ips) {
+//                logger->info("done dns-testing whatsapp endpoints");
+//                tcp_many(ips, entry, [=](Error err) {
+//                    logger->info("done tcp-testing whatsapp endpoints");
+//                    callback(entry);
+//                }, reactor, logger);
+//            }, reactor, logger);
+//        }, reactor, logger);
+//    }, reactor, logger);
```

Comment:
This commented out code can probably be removed, right?

Reply:
acked